elastic/kibana
# [Synthetics] preserve id field on monitor attributes

## Summary

Preserves monitor attribute `id` on monitor saved objects. Also preserves attribute `urls` on tcp monitors.

### Background

In https://github.com/elastic/kibana/pull/140553, we hardened our validation to ensure that unknown keys cannot be added to monitor saved objects. This will be release in 8.5.0

Prior to 8.5.0, however, the enable/disable toggle in Uptime Monitor management already introduced an unknown key: `id`. This `id` key was added to the attributes of every monitor that was enabled or disabled via the enable/disable toggle.

Because our monitor saved object contains encrypted secrets, values on the monitor saved object are added to the encryption AAD (additionally authenticated data), but default. If values are omitted from the AAD, the secrets will not be able to be decrypted. 

Because of this limitation, we must continue to keep the `id` field on all monitors until the time at which we are able to remove the id field through a saved object migration.

**Note on the `urls` field for TCP:** 
This PR also ensures that the `urls` field for tcp is preserved. The urls field was defined on all tcp monitors prior to 8.4.2 through our now removed hydration feature, which updated monitor saved objects with extra values based on the heartbeat documents. Preserving `urls` on TCP fields is necessary for the same reason it is necessary to preserve the id field: to prevent decryption errors resulting from missing fields that are part of AAD.

### Testing
1. Before checking out this pr, checkout 8.4.0 `git checkout v8.4.0`
2. Create a cluster via oblt-cli
3. Connect your local Kibana to the oblt-cli cluster by pasting in the provided kibana.dev.yml file. 
4. Update `xpack.uptime.index: remote_cluster:heartbeat-*` to `xpack.uptime.index: synthetics-*`.
5. Navigate to Uptime. Create an http, icmp, tcp, and browser monitor. (Ideally, create the browser monitor with a schedule of 2 minutes so you don't have to wait too long for multiple results)
6. Wait for hydration to occur by waiting for the `urls` field for the browser monitor to populate.
7. Disable all the monitors
8. Re-enable all the monitors
9. Check out this branch and rebuild
10. Edit each monitor
11. Ensure there are no decryption errors in your service logs. Also ensure monitors are able to be decrypted by trying to edit each monitor once again. If the monitor loads when you try to edit it, no decryption errors are present.
